### DISCUSSION TITLE
Improving the documentation structure

### DISCUSSION THREAD

**Discussion by gbishop** [UNC-Project-Open-AAC/OS-DPI#189] [General]
Using the implementation to drive the documentation is getting confusing. I'm thinking about a better organization. The **bold** names below are Wiki pages. The other names are headings within the Wiki page. **Actions** is a page describing the Actions tab and all its content (what we have now). There is no separate page for Action, Conditions, or Updates. Those could be relegated to the Terminology page. 

**Methods** will be described in one long page. If it get too long we'll break it up somehow. Likewise for **Patterns**.

What do you think?

- **Actions**
  - Action
    - Conditions
    - Updates

- **Layout**
  - **Audio**
  - **Button**
  - **Customize**
  - **Display**
  - **Gap**
  - **Grid**
    - Filter
  - **Logger**
  - **Modal Dialog**
  - **Page**
  - **Radio**
    - Option
  - **Speech**
  - **Stack**
  - **Tab Control**
    - Tab Panel
  - **Visual Scene Display**

- **Content**
  - Spreadsheet summary
  - Media

- **Cues**
  - Overlay
  - Fill
  - Circle
  - CSS

- **Methods**
  - Method
    - Timers
    - Handlers
      - Handler (Key, Pointer, Timer)
        - Conditions
        - Key Conditions
        - Responses
          - Activate
          - Next
          - Cue
          - Clear Cue
          - Start Timer
          - Emit
          - Log

- **Patterns**
  - Pattern
    - Group
    - Selector
      - Filter
      - Order by
      - Group by

**Comment by gbishop** [comment]
I have started work on this. The Patterns page is an example of what I intend.

**Comment by aetkie** [comment]
Re-visiting this discussion thread. 

I see what you've proposed for the Patterns page and how each level corresponds to components in the OS-DPI. I made some changes to the [page]( https://github.com/UNC-Project-Open-AAC/OS-DPI/wiki/Patterns ). I think there is still a need for the existing content that introduces Patterns and provides some explanation. I'll work on expanding on various levels that correspond to the OS-DPI (i.e. order by, group, etc).

**Reply by sofia-bg** [reply]
Quickly scanned through this page and this jumped out to me under  **How to Implement Patterns**

2. If you want icons to be cued from left to right and top to bottom in order the buttons appear on the screen, use Row Column

3. If you want icons to be cued by column starting from left to right on the screen, use Column Row Scan rows then columns within the row. Cycle through the columns of a row 2 times before returning to row scanning. 

    --> I am a little confused with how the row/column and column/row are explained. First I would revise to make the descriptions parallel. I might change syntax entirely to start with the name and then give the description. (i.e., **Row/Column:** cues... ). Second, my understanding is that it truly is scanning the whole row, not the individual icons. So in row column you are scanning each row from top to bottom and then when you select a row it scans left to right. Same with column/row that you scan each column from left to right and then once in a column from top to bottom. 

    --> **Cycle through the columns of a row 2 times before returning to row scanning.** this sounds like a directive, but I think it is simply explaining the way that the cue in the image is programmed, presumably you do not have to have it cycle through 2 times, you could come up with any number you want. Thinking it could sound more like explaining what is there not implying that to work it must be set to 2 times.

**5. If you want icons to be cued by control buttons (e.g., Home, Clear, Backspace, etc.) use Controls and Rows**
 
    --> I am not following what this means 

~~Also, I need a refresher, if you pick row/column or column/row is it scanning single cells at a time in those directions?~~ forgot to deleted this after I went back to try out scanning in OS-DPI mid writing this post

**Comment by aetkie** [comment]
I appreciate your feedback and a second set of eyes on this. Making edits now.

One part of Patterns that is still confusing to me is how to make the cycling start. Which component in Patterns is where you determine the key that makes the scanning start cycling?

**Reply by sofia-bg** [reply]
<img width="1512" alt="Screenshot 2023-06-06 at 4 24 15 PM" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/92817124/2fe40cc0-e847-480b-9601-e6fc3679134f">

It is under Methods. Here you see I can start the scanning by hitting either the space or the right arrow. Every time I hit either I advance down the rows, until I hit either the enter or left arrow. Once I do that then I am back to the space and right arrow to move along the cells in the row until I hit enter or left to activate. 

Although @gbishop maybe I am doing something wrong because with this current set up I can cycle through continuously? 

[grid_ex_1-3_scann.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/11668811/grid_ex_1-3_scann.zip)

**Reply by aetkie** [reply]
I am looking at this design now [grid_ex_1-3_scann.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/11668811/grid_ex_1-3_scann.zip) and I see how it scans row by row. It does scan continuously for me too, even though the cycle = 2.

My question is, how would we make the pattern do this:
1. Go row by row 
2. Go back to Row 1 and scan icon by icon.

At first I thought this would be 2 different Patterns. But you can only have 1 active at a time. Does that mean a second Group within the active pattern? has anyone done this already?

**Reply by aetkie** [reply]
Figured it out! It's controlled in Methods. Under Switch, you would have 2 Key Handler boxes to differentiate the controls for row by row and then icon by icon.

**Reply by sofia-bg** [reply]
@aetkie can you share your new design? I am not quite understanding the pattern you are describing you were aiming for, it sounds like the already existing pattern so I would like to see it to understand.

**Reply by aetkie** [reply]
It _does_ already exist. Hah. I figured it out. 

This is what I am working on.
[keyboard_predict_AE.osdpi.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/11690328/keyboard_predict_AE.osdpi.zip)

Right now, it scans through row by row. And then you can go letter by letter. 
The next thing I want to figure out is how to cycle through the Predictions row on the second sheet. But I haven't figured that out yet. Any advice?

**Reply by sofia-bg** [reply]
Not sure what you mean by on the second sheet - this might not be the exact order you want, but what I did was instead of grouping by rows, I made a new column called scan. Then I set it to order by component name. You would need to change the name of the components to whatever alphabetical order you want, but this way it does not group the predict with the first row of the keyboard. I assume this was happening because even though you don't state "row 1" in the predict sheet, it is still "row 1" for that component. 
[keyboard_predict_AE_SBG.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/11691100/keyboard_predict_AE_SBG.zip)


Also - to make my life easier I changed it so that it is right arrow to advance scans and space bar to select.

**Comment by gbishop** [comment]
Yes. It needs help for sure. I simply smashed the pages together. As (I
think) you suggested earlier, I think it (as well as Methods and Cues)
should start with examples that go from simple to complex. Grouping by rows
seems simple and (I hope) pretty obvious. Likewise grouping by columns.
Then we might point to another page that covers the Controls and Rows
example. If we want to get really fancy we could include an example where
the grouping and order is fully defined by the content from the spreadsheet.

**Reply by sofia-bg** [reply]
@aetkie and maybe my advice is wrong above, maybe instead of calling it "choosing rows columns or columns rows" and explaining how these pre-made options run, maybe we need to explain how they are both programmed. For example Row column is only named that because that is what Gary wrote into the field for Name. And then it scans rows because #row is what is entered into the field Group By. 

But can also appreciate maybe this is too complicated and we start with explaining they can choose either of the presets and how they work and then peel back to show why and how things can be changed to work differently? 

@gbishop Does anything need to actually be entered into the name field next to group by? Currently you have Rows #row deleting doesn't change anything, but does this help for logging or something else?

**Reply by aetkie** [reply]
_And then it scans rows because #row is what is entered into the field Group By._

That seems important to explain. I'll add that. 

Pre-made options are intended to get people started. But with the understanding that the scanning patterns are endless... 
I'll expand on this idea

**Reply by aetkie** [reply]
Patterns [page](https://github.com/UNC-Project-Open-AAC/OS-DPI/wiki/Patterns) updated. That's it for now. Will continue with further iterations as additional UIs (and my own understanding) progress.

**Comment by aetkie** [comment]
I think Sofia's working UI (when complete) will be a great example of scanning with controls and rows. 

 _If we want to get really fancy we could include an example where
the grouping and order is fully defined by the content from the spreadsheet._ 

You would have to walk me through this a bit...

**Reply by gbishop** [reply]
Imagine you have an extra field (or 2 or 3) in your content. Then you can Group By or Order By those fields. So if I wanted to scan right to left for my Hebrew communication board I could make that happen with a (say) Order column in my content that counts down instead of up. Or I could create a spiral scan by having the order column count up in the order I wanted the buttons visited. Maybe you're doing a keyboard and want the most frequently used letters near the center of the screen and want to scan in concentric circles; just assign the appropriate number to each letter and then Order By that. 

I'm not suggesting that any of those is a good idea. Just illustrating what could be done.

Does that help?

**Reply by sofia-bg** [reply]
@aetkie the current prototype I am working on does rely on the spreadsheet to define what data I want pulled in, and I have it pre-defined so I can play around with data in different areas of the UI (quadrants, halves, thirds, etc.) and doing it that way makes it so that I don't really have to change much in the OS-DPI except for which column in my spreadsheet I want to use to declare with data to fill in... 

All that to say I am not actually using scanning in that design, its all point and dwell. But I am happy to think about ways to use the spreadsheet so you don't have to program as much on the OS-DPI side as we think about example we might want to make.

**Reply by aetkie** [reply]
@gbishop that makes sense. I can envision how that spreadsheet would be organized and reflected in the design.

**Reply by aetkie** [reply]
@sofia-bg I downloaded the spreadsheet for the design you're working on (block_to_direct). Maybe I can use the content you have to simulate different scanning patterns, other than row/column column/row. Let me poke around this week and see what I can come up wiht

**Comment by gbishop** [comment]
I'm pretty sure you could do it with the same controls by adding another
group to the pattern. You'd move through the first group row by row (and if
you selected a row, then button by button on that row) then when (if) you
exhausted that group you'd move to the second group where you'd visit the
items sequentially. I can't imagine why you'd want that behavior but I
think you could get it.

**Reply by aetkie** [reply]
<img width="593" alt="Screenshot 2023-06-08 at 12 30 52 PM" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/113473933/7a4f4dec-bbc0-4199-befe-e5ab310f5afb">

this is the intended scanning pattern. Rows 1 through 4, and then up to Row 5 (the word prediction). I'm thinking that as the user is inputing letters, eventually the word that they want will appear in Row 5 and there needs to be  a pattern to get up there. The prediction content is on a separate sheet from the keyboard. Is this making sense?

**Reply by gbishop** [reply]
I have added several patterns to your design: 

[keyboard_predict_AE_pattern_examples.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/11692916/keyboard_predict_AE_pattern_examples.zip)

Several things I would like you to note.
1. "Screen order" is the same as "None" above. It simply visits the keys in order. 
2. "Component order" creates a group out of each component but once you select the component you visit each button in order. Note that a group with one member (the spacebar) does not require an additional selection; it is as if the group wasn't there.
3. "Letters then prediction" creates no groups. It simply visits the keyboard in order, then the spacebar and finally the prediction. The key to notice is that a Selector does not create a group (unless you use GroupBy). When you come to the end of the entries in a Selector you simply go to the next one. 
4. "Letters with rows then prediction" groups the kb into rows, followed by the spacebar, then the predictions. So we have a group for each row but still the selectors pass from one to another.
5. "Letters with rows then prediction group" wraps the predictions in a group so I don't have to go through all of them to get back to the kb.
6. "Crazy Town" attempts to show what you can do by annotating your content.

I'm not recommending any of these but I thought it might be helpful to have some examples to consider.

**Reply by gbishop** [reply]
One more interesting bit I forgot yesterday. Empty groups are ignored. So, for example, if you arranged for the predict grid to be empty when the prefix is empty, your scanner won't stop on that grid when it isn't useful.

**Reply by sofia-bg** [reply]
Trying to catch up - so it looks like from what you have written that the update I made to Andrea's worked just because I made the "scan" in the spreadsheet? 

@aetkie the design I made by adding scan does that order that you are wanting EXCEPT that then it starts back up at the first row of QWERTY after you select a letter. So I am still wondering if maybe you would want it to start on the predict row? Otherwise if you select a letter then you have to scan the entire keyboard before you could go to a prediction? 

@gbishop or could she add another pattern with a different key that when struck would take you directly to the predict grid and override the other pattern?

**Reply by sofia-bg** [reply]
@gbishop ah of course asking questions then looking. I couldn't figure out how to do filter yesterday because I was expecting the same sets of boxes to pop up that pop up when you add filter in the display tab. with only one box I wasn't sure what to do.

**Reply by sofia-bg** [reply]
Last comment for the morning, this is interesting to see the ways we attack problems differently that can make documentation hard. My go-to is to lean on the spreadsheet first and make new columns that I then can control whereas here Gary manipulated the OS-DPI without having to add anymore data.

**Reply by gbishop** [reply]
Why isn't this filter the same as the Grid filter? I'll have to think about that. 

Also, I'd rather see you add to your spreadsheet than write complicated Patterns but I wanted to demonstrate the types of things you can do with Patterns.

**Reply by aetkie** [reply]
@sofia-bg I agree. There are many ways to solve the same problem. Like Gary, I want to make the OS-DPI do the work for me. My brain is almost never going back to the spreadsheet. But I'm glad yours does. Also these long discussion threads are very helpful for documentation.

**Reply by aetkie** [reply]
@gbishop your example patterns were extremely helpful. Thank you!

**Comment by sofia-bg** [comment]
When you add filter in layout: 
<img width="755" alt="Screenshot 2023-06-09 at 10 39 32 AM" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/92817124/4b168751-2ba3-44fb-9fa7-6c3f2718dc6b">

When you add in patterns: 
<img width="754" alt="Screenshot 2023-06-09 at 10 39 51 AM" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/92817124/2fbbdaa9-5c4d-4908-9170-de5cfbe10abf">

I just wasn't sure what to write in with only the one box.

**Comment by gbishop** [comment]
I understand. I'm wondering why I made them different? I'll have to try to
figure out why past me did that.
